### PR TITLE
Added DNNL cache management for DyGraph

### PR DESCRIPTION
**Pull request by arlesniak**
### PR types
New features

### PR changes
Others

### Describe
Added DNNL cache management for DyGraph for mnist, resnet, mobilenet and topologies from models repo

### CODE REVIEW

**Review by jczaja — DISMISSED** [review]
LGTM

**Review by sfraczek — DISMISSED** [review]
LGTM

**Review by grygielski — DISMISSED** [review]
LGTM

**Comment by jczaja** [comment]
@luotao1 Could you please start your review?

**Comment by luotao1** [comment]
LGTM @phlrain Could you have a review?

File: paddle/fluid/imperative/layer.cc
Comment (phlrain) [inline paddle/fluid/imperative/layer.cc:200]: ClearGradient 这个操作会在每一个step完成之后，所有的parameter都会执行，这个ClearMKLDNNCache操作的性能是什么样子的，如果很快，这个地方就没有问题，如果性能不是很高，这个位置可能需要优化下位置

Comment (luotao1) [reply paddle/fluid/imperative/layer.cc:200]: Translate from @phlrain:
`ClearGradient` will be run after each setup, and all parameters will run this function. Thus, 
- how about the performance of `ClearMKLDNNCache`?
- `ClearMKLDNNCache` will be run in NON-MKLDNN dygraph model, Could you add some `if-else` to skip the NON-MKLDNN dygraph model? i.e
```
#ifdef PADDLE_WITH_MKLDNN
if(FLAGS_use_mkldnn)
  ClearMKLDNNCache(grad_t->place());
#endif
```

Comment (arlesniak) [reply paddle/fluid/imperative/layer.cc:200]: Thank you for the comments. Taking into consideration DNNL cache use in DyGraph  explained in [25466](https://github.com/PaddlePaddle/Paddle/issues/25466) we need to have a place in flow (i.e. Layer.clear_gradients) where we can clear DNNL cache after iteration forward->backward->update cycle. The advantage of Layer.clear_gradients() is clear meaning in the iteration flow.
- ClearMKLDNNCache will be run on Layer.clear_gradients()->*clear_gradient() calls. Performance of clearing the cache map underneath will depend on the size of map so the first call will be longer
- Thank you for suggestion I used the FLAGS.

**Comment by arlesniak** [comment]
For some reason the CI fails on test_parallel_executor_seresnext_with_reduce_gpu due to accuracy diff. I'm not sure if there is anything related to it, what has been proposed in this PR. I've rerun it.

File: paddle/fluid/common_flags.cc
Comment (luotao1) [inline paddle/fluid/common_flags.cc]: Why not put into `paddle/fluid/platform/flags.cc`?

**Review by luotao1 — COMMENTED** [review]
How about split this PR into two PRs for merging faster?
- No not change `paddle/fluid/imperative/layer.cc`, i.e, use `ClearMKLDNNCache` function and update other files.
- Then, use `ClearMKLDNNCache` in `paddle/fluid/imperative/layer.cc`. You may change `paddle/fluid/platform/flags.cc` at the same time.

**Comment by luotao1** [comment]
`test_parallel_executor_seresnext_with_reduce_gpu` is random fail, you can restart it.

File: paddle/fluid/common_flags.cc
Comment (arlesniak) [reply paddle/fluid/common_flags.cc]: The idea looked nice but the MAC CIs (MAC_PR_CI, MAC_PR_CI_python35) and  PR_CI_Py35 found the following error:

> [15:45:03]W:	 [Step 1/1] Undefined symbols for architecture x86_64:
[15:45:03]W:	 [Step 1/1]   "fLB::FLAGS_use_mkldnn", referenced from:
[15:45:03]W:	 [Step 1/1] ld: symbol(s) not found for architecture x86_64
[15:45:03]W:	 [Step 1/1] clang: error: linker command failed with exit code 1 (use -v to see invocation)
logs: [PR_CI_Py35](https://paddle-docker-tar.cdn.bcebos.com/buildLog/379449.log) [PaddlePaddle_MAC_CI](https://paddle-docker-tar.cdn.bcebos.com/buildLog/379450.log) [PaddlePaddle_MAC_CI](https://paddle-docker-tar.cdn.bcebos.com/buildLog/379449.log)

So I've brought back the original idea to put 'use_mkldnn' at the top level CMake to be accessible for platform and imperative as well.

**Comment by arlesniak** [comment]
The idea of making 2 PR's looks nice however please keep in mind that we need the clearing code in several places (2 places in layer.cc file) that's why we brought the changes and refactoring together. We are open for suggestions and we can make 2 PRs if it is beneficial for you.

**Comment by sfraczek** [comment]
We can split the PR next week if you want

**Comment by arlesniak** [comment]
I'm going on vacation and @sfraczek will take over.

**Comment by luotao1** [comment]
> We can split the PR next week if you want

@sfraczek You can wait for the review from @phlrain at first.
@phlrain Could you help review again?

**Review by phlrain — APPROVED** [review]
LGTM

**Pull request merged**
